### PR TITLE
[READY] Battery overhaul 2 electric boogaloo

### PR DESCRIPTION
**Pull request by Inglonias**
#### Summary
```SUMMARY: Balance "Overhauls how the game handles batteries"```

#### Purpose of change
This is a complete overhaul of how battery powered tools are handled, with the intent of making that aspect of the game less abstract and more realistic.

Implements #29536

#### Describe the solution
All batteries are now magazine-style, and all tools that formerly accepted batteries now accept these magazines instead. You cannot unload or reload these batteries directly, but they do have the RECHARGE flag. There are 3.5 different categories of batteries, and they're detailed in the compatibility chart below:
image

**Ultralight:** Think of these as coin cell batteries. Used where size is important or the manufacturers are cheap bastards (which is why it's standard in the talking doll, for instance). Note that while tools in this category have batteries that are compatible with light tools, light batteries cannot fit into ultralight devices.

**Light:** Think of this like a smartphone battery. The high-capacity variant is a high-end one for consumer goods that need a bit more juice like professional camera equipment. Note that tools in this category can fit ultralight batteries.

**Medium:** These are equivalent to the batteries you'll find on modern cordless tools. The high capacity ones are for things like military-grade equipment (mess kits) and fancier electronics (L-sticks)

**Heavy:** These are car batteries in all but name. The high capacity heavy batteries currently only come standard on a UPS

#### Describe alternatives you've considered
The battery system as-is works well enough, but it's pretty abstract and has needed a serious upgrade for a while. Since electrical tools can already accept battery magazines no problem, this is the logical conclusion. This solution involves no new C++ code, only JSON.

#### Additional context

I made a light balance pass on any existing tools in the base game. I tried to keep things about the same in terms of overall performance, but this probably needs balance work before it can be merged. It also needs testing. If you find any places where batteries still spawn, let me know. There's no real rush on this PR, and I suspect that fully removing the old battery spawns will take a while.

**What will happen to my existing games and any batteries I may have?**
Good question! As far as I'm aware, you'll keep the charges and magazines will sprout.

Still, I realize this might be inconvenient for those of you with long-running games. For those, magazine-style batteries can be reloaded (but not unloaded). This will likely go away a few weeks after this PR is merged.

### CODE REVIEW

**Comment by Fris0uman** [comment]
- How does one recharge regular batteries ?
 - What's the difference between UPS and regular battery in this system ?

**Comment by darktoes** [comment]
I like the sound of this, seems pretty well thought out, though it sounds like light and light minus type batteries shouldn't have the RECHARGE flag. I think they should remain the "disposable" type batteries. There's also the possibility of a disposable variant of the medium size battery, to model D/9V/Lantern batteries.

Sounds like the legacy batteries should still be reloadable by default, so those with ongoing games will be able to use up their legacy batteries without requiring the mod for a while.

**Comment by Amoebka** [comment]
So, let's say I have a soldering iron and three batteries (assuming it uses medium ones, irrelevant here) for it, medium(13), medium(24) and medium(7). Will I be able to craft a recipe that needs soldering iron (40 charges)?

**Comment by ifreund** [comment]
> So, let's say I have a soldering iron and three batteries (assuming it uses medium ones, irrelevant here) for it, medium(13), medium(24) and medium(7). Will I be able to craft a recipe that needs soldering iron (40 charges) under the current implementation? Logically, I should be able to, and preferably without manually clicking the tool reload button twice.

Incremental change consumption is coming up next in #29210 and will make that possible.

**Comment by Inglonias** [comment]
> Sounds like the legacy batteries should still be reloadable by default, so those with ongoing games will be able to use up their legacy batteries without requiring the mod for a while.

This is incorrect. You will need the mod immediately to use your legacy batteries.

**Comment by Inglonias** [comment]
> How does one recharge regular batteries ?
...I don't actually know what the recharge flag does.
> What's the difference between UPS and regular battery in this system ?
I guess the UPS is just a really fancy extra battery pack in this new system. Isn't that what it was before anyhow?

**Comment by darktoes** [comment]
> This is incorrect. You will need the mod immediately to use your legacy batteries.

I'm saying you should make it that way. People shouldn't be able to obtain more legacy batteries, but anyone with an ongoing save will likely have a bunch in storage and lots more lying around in houses and such. If you make it so the batteries can be loaded but not unloaded, then it can't be exploited, and anyone with an existing save is covered without needing to download a mod.

**Comment by Human-Shield** [comment]
Magazines are a good concept. It would be great if "charge units" were real world measures like kwh and scale with vehicle power. Does the car battery mod still work for tools? 

Need a way to charge without a UPS, should be doable with copper wires if you need to transfer charges over time between magazines (with % lost).

**Comment by Inglonias** [comment]
Car battery mod should still work, yes.

One unit of battery already represents 1 kJ of energy.

Charging without a ups might require new C++ code, which I don't think is in scope.

**Comment by thethunderhawk** [comment]
Sounds like we could kill two birds with one stone by reworking UPS-modded tools to just take power from a central battery, plugged into a UPS unit. Maybe the UPS could accept any type of battery, including small car batteries or whatever. That’d be the benefit of UPS-modding your tools, simplicity.

Meanwhile the current UPS recharging station could be reworked into a generic battery charging station, common in most vehicles, easily craftable with low electronics skill, whose only role is to charge batteries of all types.  

Obviously that’s out of scope though. This is a good step in the right direction, I’m excited about it.

**Comment by Inglonias** [comment]
I'm not going to merge this PR without some way to recharge these battery packs. That doesn't feel fair. A good short-term fix would be to allow the UPS charging station in vehicles to recharge these batteries, like you said. (They may already do this but I've been too busy to do more than basic testing lately so I'm not certain.)

In the long term, there definitely needs to be a way to consolidate battery charge besides the vehicle charging station, though.

**Comment by Fris0uman** [comment]
UniversalPowerSuplly (I think ...) was the only rechargeable battery, it was recharged in a UPS recharging station or by loading batteries in it and would power UPS compatible tools that generally only accept UPS. It's also the power source of power armors and some fancy guns.

From what I get this PR makes every battery a different kind of UPS. So UPS shoul either desappear or be reworked into something else, maybe following thethunderhawk suggestion.

**Comment by thethunderhawk** [comment]
Personally I think the vehicle charging station will work as the primary method for charging batteries until someone makes power grids a thing. Right out of the box players will have the ability to hook up portable generators, dynamos, and solar panels to their battery chargers. I also think there’s room for the solar backpacks to have a role here, plus maybe a solar mod for very small tools, and a mod for car batteries that allows them to in turn charge up smaller batteries.

**Comment by Fris0uman** [comment]
Also what happens whith the battery system cbm ? A CBM that allows you to use regular battery to get power.

**Comment by tinukedaya** [comment]
I'm not too familiar with the Bright Nights mod from Coolthulhu, but if I remember correctly he added some rechargeable batteries and changed the UPS recharge station to generic recharge station. Might be worth checking? 

Otherwise I can't wait to see this happening!

**Comment by Inglonias** [comment]
> Also what happens whith the battery system cbm ? A CBM that allows you to use regular battery to get power.

That's something else I'll need to test. Thanks for bringing it to my attention.

**Comment by FuelType-Memes** [comment]
Is it possible to make some appliances take batteries as a magazine, but actually recharge them? So that way you can slot "reload" you solar backpack/dynamo/plutonium charger/whatever and it will charge over time, instead of drawing energy?

**Comment by jeremyshannon** [comment]
I'm gonna further support darktoes' suggestion here by pointing out that some players will have trouble with the idea of adding a mod just so their battery stockpiles aren't invalidated. It's not like the game offers an easy way to add mods to an existing save, you have to go in with a text editor, and that scares some people.

**Comment by Inglonias** [comment]
> I'm gonna further support darktoes' suggestion here by pointing out that some players will have trouble with the idea of adding a mod just so their battery stockpiles aren't invalidated. It's not like the game offers an easy way to add mods to an existing save, you have to go in with a text editor, and that scares some people.

Alright. You've convinced me. I just removed the NO_RELOAD flag from the batteries.

I also did some more testing.

> I'm not too familiar with the Bright Nights mod from Coolthulhu, but if I remember correctly he added some rechargeable batteries and changed the UPS recharge station to generic recharge station. Might be worth checking?

UPS charging stations charge anything with a RECHARGE flag. Batteries have that flag now, so they can be recharged.

> Also what happens whith the battery system cbm ? A CBM that allows you to use regular battery to get power.

Testing shows that you drink the charge out of the battery and leave the magazine empty.

**Comment by Inglonias** [comment]
Ah, dang it.

Don't merge yet. I failed to adequately resolve the merge conflicts and batteries will still spawn in some locations.

**Comment by Inglonias** [comment]
Alright. Merge conflicts have been resolved properly now, it looks like.

This could probably be merged as-is and not break too much now, but obviously I'd like to avoid that if possible. I'll also be unable to work on code this weekend, so maybe avoid merging until I return on Monday at the earliest.

**Comment by paulenka-aleh** [comment]
Just out of curiosity, will there be in game in future convencional non-rechargable batteries like ones I have in my flashlight, TV remote, remote keyboard, weather stations? In-game items I would give such batteries would be radios, flashlights, RC controls, radiation biomonitors, geiger counters, probably some of two-way radios and some cameras (not pro). In compare to rechargable ones they have higher capacity but aren't designed to be recharged and thus shouln't be.

Also it feels a bit odd to drink li-ion battery eith battery system in an instant without blowing it up but it's a whole other story

**Comment by Inglonias** [comment]
> Just out of curiosity, will there be in game in future convencional non-rechargable batteries like ones I have in my flashlight, TV remote, remote keyboard, weather stations? In-game items I would give such batteries would be radios, flashlights, RC controls, radiation biomonitors, geiger counters, probably some of two-way radios and some cameras (not pro). In compare to rechargable ones they have higher capacity but aren't designed to be recharged and thus shouln't be.

I wasn't planning on it, but those should be trivial to add if there's demand for it. Just need to add it to the item entries and existing item groups.

**Comment by Inglonias** [comment]
I've just looked at the Travis builds and found some crafting_test failures, which concern me. I'm going to have to look into them and see if they're just based on numbers I've changed in my balance pass, or if charges are not being used in tools properly.

**Comment by Inglonias** [comment]
Looks like there's something up with the test's numbers. Not sure what, but in my own tests, tools are used properly.

**Comment by echomirage1** [comment]
With this system every place where people used batteries has to be filled with chargers (like the chargers of cordless drill batteries in real life). Even if there is no way now to connect the charger to the power network, they could be disassembled for electronic junk or be able to be mounted as vehicle chargers.

**Comment by Inglonias** [comment]
Yes, I plan to add the recharging station vehicle part to several item groups as part of this PR. First I need to fix the crafting tests, though.

File: tests/player_helpers.cpp [hunk 1]
```diff
@@ -18,7 +18,7 @@ int get_remaining_charges( const std::string &tool_id )
     } );
     int remaining_charges = 0;
     for( const item *instance : items ) {
-        remaining_charges += instance->charges;
+        remaining_charges += instance->ammo_remaining();
```
Comment (Inglonias) [inline tests/player_helpers.cpp]: This was all it took to get crafting tests to pass locally.

**Comment by Inglonias** [comment]
I believe this PR is ready now, pending the automated tests. I was able to get automated tests to pass locally after a great deal of trial and error, and all merge conflicts are resolved. I've added the recharge_station item to a few more item_groups, so it should spawn a little more often.

**Comment by Inglonias** [comment]
Scratch that. If you craft something, it comes with a free full battery.

File: data/json/recipes/recipe_medsandchemicals.json [hunk 2]
```diff
@@ -57,7 +57,17 @@
...
-    "components": [ [ [ "battery", 125 ] ] ]
+    "components": [
+      [
+        [ "light_minus_battery_cell", 20 ],
+        [ "light_battery_cell", 10 ],
+        [ "light_plus_battery_cell", 5 ],
+        [ "medium_battery_cell", 2 ],
+        [ "medium_plus_battery_cell", 2 ],
+        [ "heavy_battery_cell", 1 ],
+        [ "heavy_plus_battery_cell", 1 ]
```
Comment (Night-Pryanik) [inline data/json/recipes/recipe_medsandchemicals.json]: So these battery cells are all acid-based? Not Li-Ion, not Li-Poly?

Comment (Inglonias) [reply data/json/recipes/recipe_medsandchemicals.json]: Um. I dunno. They're rechargeable, but I wanted to keep as much compatibility with existing recipes as possible.

**Comment by Inglonias** [comment]
I seem to have fixed the free batteries coming with crafting, but at the cost of things not spawning with batteries inside mapgen anymore, which kind of sucks.

**Comment by Inglonias** [comment]
Alright. This is the deal, currently.

Crafted items do not spawn with batteries.

Monsters that drop battery powered tools seem to be able to either drop them with a full battery, or no battery at all. There is no in-between.

Currently, things spawned from itemgroups DO NOT spawn with batteries.

File: data/json/recipes/recipe_medsandchemicals.json (hunk 2 above)
Comment (Night-Pryanik) [reply data/json/recipes/recipe_medsandchemicals.json]: I'd remove the recipe altogether. It always looked weird to me. But that's only my opinion.

Comment (Inglonias) [reply data/json/recipes/recipe_medsandchemicals.json]: As far as I know, there's only two acid recipes and I'm not sure if the other is possible without acid rain. I'm keeping this for now so people can craft acid at all, but later I'll add some non-rechargeable batteries that are acid based, and move the recipes over.

**Comment by Inglonias** [comment]
Updated the original post in this conversation with new data.

I don't plan to change things around anymore. This PR is done as far as I'm concerned. That said, I don't know for sure what the current behavior of existing tools in existing games will be once this PR gets merged, so I will be testing that when I get home today.

I've started a test game and I'm not noticing too much of a problem on Day 2 so far. I'm having trouble finding charged batteries, but I've only explored a few houses so far and it's not actually hindering anything I'm doing yet.

**Comment by Inglonias** [comment]
This PR is as ready as it's ever going to be, with one possible exception.

I still haven't adjusted any battery powered tools outside of the base game (so stuff in Aftershock, Bright Nights, etc.), nor have I altered any non-base game item groups or mapgen. Should I do that, or should I leave that for another PR/let someone else do it?

**Comment by Inglonias** [comment]
Professions do not start with charged electronic devices, and I cannot seem to make them do so.

Is this a problem?

**Comment by Inglonias** [comment]
The longer this goes without being merged, the more I'm gonna chew on it. Obviously, if things aren't linted properly, or tests are failing, that takes top priority, but still... Is there anything else missing from this PR that I haven't covered?

* Item spawns replaced
* Migration is as easy as it will get
* Map gen is replaced
* Recipes are replaced
* Mods are updated (though I didn't rebalance them)
* Professions are updated as best I can
* All relevant items are updated

**Comment by Night-Pryanik** [comment]
It's just lack of time to test and merge.

**Comment by Vasyan2006** [comment]
>Ultralight: Think of these as coin cell batteries.

C type battery (1.5V@8000mAh) has capacity 43.2 charges, they are slightly larger than common coin batteries.

>Light: Think of this like a smartphone battery.

Common smartphone power cell (3.7V@2500mAh) can hold only 33.3 charges. Setting their capacity to 150 is a kind of overestimation.

**Comment by Inglonias** [comment]
I'm not any sort of battery expert. I just wanted batteries to make slightly more sense than they do right now. As such, I have no justification here in terms of realism for my numbers. I just didn't want to rock the boat too hard in terms of the numbers in the game, since this is already a pretty major change. If you have more sensible numbers in mind, I'd love to hear them, but I'm not the person to come up with those numbers.

Cell phones and smart phones in the game can hold 120 charges when full. Correctly or not, I'm using the game's current values as a benchmark for my own whenever possible. I justified making the high capacity version hold 150 by saying that a 120 capacity battery really isn't much better than a 100 capacity battery. So it's a gameplay justification, not realism.

If you've got better numbers than I do, again, please share them. I'd be happy to try things out with your numbers to see what happens.

**Comment by Vasyan2006** [comment]
Charge capacity for items was set before the introductions of 1 charge = 1 kJ rule and it can be very arbitrary. Imho based on irl batteries for different items this chart should be:
Ultralight - mp3 players and game watch: 10 charges
Light - smartphone - 30 charges,
Medium (2x NiMH C battery) - flashlight - 45 charges,
Heavy (2x NiMH D battery) - HD flashlight - 90 charges, 
Heavy+ - cordless drill - NiMH batteries - up to 110 charges from this link
https://www.amazon.com/Battery-Packs-Drill-Chargers/s?rh=n%3A552354%2Cp_n_feature_keywords_browse-bin%3A5909259011

Nonchargeable alkaline batteries have much higher capacity compare to Li-Ion or NiMH, but they are obviously single use items. 40 charges in the last post was for alkaline C battery. 

Hotplates, jackhammers, kilns and forges can easily consume 1 kW (6 charges every turn) of power and should use vehicle storage batteries with preinstalled vehicle adapter.

UPS is crafted from the small storage battery and should have the same capacity.

This change can make many items unusable until their power consumption will be recalculated in Joules. So maintainers must decide about implementing these changes now or in separate PR.

**Comment by Zireael07** [comment]
Good idea @Vasyan2006 but IMHO for a separate PR. This one is plenty big already.

**Comment by Inglonias** [comment]
I agree. Those are definitely things to look at for a rebalance later. Not for this PR, though.

**Comment by paulenka-aleh** [comment]
If smartphones are going to have 30 charges their battery consumption should be revised too. I have 2500mAh battery and able to surf internet for 4 hours without caring about power saving. In-game cellphone on-screen flashlight uses up 10 charges every 30 minutes.

I suppose it's valid for all the other tools.

**Comment by thiagoshima** [comment]
> Charge capacity for items was set before the introductions of 1 charge = 1 kJ rule and it can be very arbitrary. Imho based on irl batteries for different items this chart should be:
> Ultralight - mp3 players and game watch: 10 charges
> Light - smartphone - 30 charges,
> Medium (2x NiMH C battery) - flashlight - 45 charges,
> Heavy (2x NiMH D battery) - HD flashlight - 90 charges,
> Heavy+ - cordless drill - NiMH batteries - up to 110 charges from this link
> https://www.amazon.com/Battery-Packs-Drill-Chargers/s?rh=n%3A552354%2Cp_n_feature_keywords_browse-bin%3A5909259011
> 
> Nonchargeable alkaline batteries have much higher capacity compare to Li-Ion or NiMH, but they are obviously single use items. 40 charges in the last post was for alkaline C battery.
> 
> Hotplates, jackhammers, kilns and forges can easily consume 1 kW (6 charges every turn) of power and should use vehicle storage batteries with preinstalled vehicle adapter.
> 
> UPS is crafted from the small storage battery and should have the same capacity.
> 
> This change can make many items unusable until their power consumption will be recalculated in Joules. So maintainers must decide about implementing these changes now or in separate PR.

One thing to note is that the game happens in 2050, from now until them the battery tech could have advanced.

**Comment by Inglonias** [comment]
> Charge capacity for items was set before the introductions of 1 charge = 1 kJ rule and it can be very arbitrary. Imho based on irl batteries for different items this chart should be:
> Ultralight - mp3 players and game watch: 10 charges
> Light - smartphone - 30 charges,
> Medium (2x NiMH C battery) - flashlight - 45 charges,
> Heavy (2x NiMH D battery) - HD flashlight - 90 charges,
> Heavy+ - cordless drill - NiMH batteries - up to 110 charges from this link
> https://www.amazon.com/Battery-Packs-Drill-Chargers/s?rh=n%3A552354%2Cp_n_feature_keywords_browse-bin%3A5909259011
> Nonchargeable alkaline batteries have much higher capacity compare to Li-Ion or NiMH, but they are obviously single use items. 40 charges in the last post was for alkaline C battery.
> Hotplates, jackhammers, kilns and forges can easily consume 1 kW (6 charges every turn) of power and should use vehicle storage batteries with preinstalled vehicle adapter.
> UPS is crafted from the small storage battery and should have the same capacity.
> This change can make many items unusable until their power consumption will be recalculated in Joules. So maintainers must decide about implementing these changes now or in separate PR.
> 
> One thing to note is that the game happens in 2050, from now until them the battery tech could have advanced.

I'm not going to use that as an argument. Currently, batteries are balanced around lighter tools having roughly 100 charges and I'm not going to mess with that in this PR. If someone wants to implement Vasyan's numbers later in a balance pass, great. I'll happily help them if they need it. For now, I'm not going to affect balance that much if I can avoid it.

**Comment by Night-Pryanik** [comment]
> One thing to note is that the game happens in 2050

No, it's not. Actually it's present time, maybe plus-minus a few years.

**Comment by adamkad1** [comment]
plutonium cells stay as normal?

**Comment by egorzaderiy** [comment]
UPS mod can't be attached to pre-patch tools. Is this deserves issue?

**Comment by l29ah** [comment]
Ok now my survivor headlamp doesn't consume batteries but works, cool.

**Comment by adamkad1** [comment]
you know, this should probally be a separate mod untill everything is done, because its way more work than they think it is

**Comment by Night-Pryanik** [comment]
It's not a good idea for project to present feature that will be mainlined eventually in form of a mod. People might not be wanting to activate this mod, so we don't get a feedback for new mechanics.
For those who don't want to test new mechanics, there is a mod that reverts this change back to old batteries.

**Comment by adamkad1** [comment]
there is?

**Comment by Night-Pryanik** [comment]
> there is? where?

https://github.com/CleverRaven/Cataclysm-DDA/tree/master/data/mods/Battery_Migration

**Comment by Inglonias** [comment]
> also on opinion side, i think people who willingly take things that are unfinished are more effective at finding errors than random people

There was not one but TWO threads on Reddit asking for review. If you wanted to test it, you had every opportunuty.

**Comment by adamkad1** [comment]
> > also on opinion side, i think people who willingly take things that are unfinished are more effective at finding errors than random people
> 
> There was not one but TWO threads on Reddit asking for review. If you wanted to test it, you had every opportunuty.

do i look like a person who uses reddit?

**Comment by adamkad1** [comment]
> > there is? where?
> 
> https://github.com/CleverRaven/Cataclysm-DDA/tree/master/data/mods/Battery_Migration

thats just reloading batteries with legacy cells, doesnt help for stuff that pop errors because invalid mag. guess ill hang off playing the game

**Comment by kevingranade** [comment]
> do i look like a person who uses reddit?

Entitlement level and bad idea quotient checks out, yea.

**Comment by kevingranade** [comment]
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/how-to-recharge-the-new-battery-cells/20081/2

**Comment by I-am-Erk** [comment]
Hey there. I'm working on plutonium cells and came across the "atomic" battery. What is this meant to be? It shows up in a lot of places and has limited charges, which doesn't seem to make sense with atomic.

**Pull request merged**